### ISSUE TITLE
Enhance the filter "sliding_window_moving_average" for smoothing over a given time/duration

### ISSUE THREAD

**Issue by Photelegy** [dbuezas/lovelace-plotly-graph-card#248]
**Describe the solution you'd like**
The filter `sliding_window_moving_average:` smoothes the graph by a number of datapoints (in the parameter `window_size:`) in the shown window. (https://github.com/dbuezas/lovelace-plotly-graph-card#offsets)
It would be very helpful to smooth by time-duration.
(With `lambda:` I had this functionality as shown. But lambda is now removed in v3.0.0)

```yaml
    lambda: |-
      (ys, xs) => {
        const MINUTES = 30               // <-- change here
        const window_size_ms = 1000*60*MINUTES 
        const result = {x:[], y:[]}
        let x_next = +xs[0] + window_size_ms
        let y_accum = 0
        let y_count = 0
        for (let i = 0; i < xs.length; i++) {
          const x = +xs[i]
          const y = +ys[i]
          if (x>x_next){
            result.y.push(Math.round(y_accum/y_count * 10) / 10)
            result.x.push(new Date(x_next))
            y_accum = 0
            y_count = 0
            x_next = +x + window_size_ms
          }
          if (y){
            // ignore unavailable/unknown state points
            y_accum += y
            y_count ++
          }
        } 
        return result;
      }
```

The filter `sliding_window_moving_average:` at the moment looks as followed:
```yaml
    filters:
      - sliding_window_moving_average:
          window_size: 1
          extended: true
          centered: false
```

### So the only difference would be, that it smoothes over a time-duration if there is a time-suffix. If there's no time-suffix it just works as it works now.

**How would it be defined in yaml?**
Maybe it could be included in the `parameter window_size:`
```yaml
    filters:
      - sliding_window_moving_average:
          window_size: 30m        # or 45s, 2d, ...
          extended: true
          centered: false
```
(With the time format like it's used in https://github.com/dbuezas/lovelace-plotly-graph-card#duration)

**Additional context**
Just if more information is neede, here my first Discussion (question) about this: https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/234

The following is a example of how I think the whole card could look as yaml:
```yaml
type: custom:plotly-graph
entities:
  - entity: sensor.plugmeter1
    name: Voltage
  - entity: sensor.plugmeter2
    name: Voltage
    filters:
      - sliding_window_moving_average:
          window_size: 10m            # example for an individual entity
          extended: true
          centered: false
hours_to_show: 24
refresh_interval: 10
layout:
  height: 200
  margin:
    t: 50
    l: 50
    r: 50
  yaxis:
    fixedrange: false
    range:
      - 220
      - 240
    title: Voltage [V]
  showlegend: false
defaults:
  entity:
    filters:
      - sliding_window_moving_average:
          window_size: 30s            # example for the defaults (if not individually defined)
          extended: true
          centered: false
    show_value: true
    line:
      shape: line
```
It's nearly the same as now, just with the time-suffix.

**Comment by dbuezas** [comment]
This would be really good indeed. For consistency I should probably add this option to `median` too.
One difference between this and the lambda implementation you shared above is that the sliding window average outputs as many datapoints as there are in the input (at least if `extended: true`).
I'll definitely give this a try, but if you need this urgently, remember that you can use the `fn:` filter, which is pretty much the same as a lambda, except that you have to return `{xs, ys}` instead of `{y,x}.

**Comment by dbuezas** [comment]
Ok I took a look at this. Having a time window creates issues.
Although it is implemented in a more efficient manner (and the `extended` mode makes the window size dynamically grow and shrink on the left and right extremes), the concept behind the sliding window averaging algorithm is this:
* there is an index `i` represents the end of the window.
* `i` starts at `window_size`
* the average of all datapoints between ` i - window_size` and `i` is computed, those are the smoothed output datapoints
* `i` is increased by one unit until it reaches the last datapoint.


Moving to a time based window, means that when the index increase, multiple datapoints on the left of the window may need to be removed from the window. This results in the possibility of the steps on the x axis getting jumpy, and multiple corner cases needing consideration when implementing the `extended` mode.

When algorithms that do these kinds of basic mathematical operations start getting messy, it usually means that.... they stopped doing basic mathematical operations. They won't always return exactly what one expects (for example, there will be less output than input datapoints, which is unexpected)

Therefore, before I put too much time on this I suggest we try a different approach: force datapoint indexes to actually represent time. e.g, if there are exactly 1 datapoint per minute a window_size of 10 means the same as a time_window 10 minutes.

And we have a filter that does that already! `resample: 1m`.

I think that the only circumstances under which the 2 yamls below will differ, is when a sensor is outputting data in a very irregular manner, and in that case, a time window would have the slightly weird behaviour I described above.

```yaml
filters:
  - sliding_window_moving_average:
      window_size: 10m # <---- 10 minutes!
```

and

```yaml
filters:
  - resample: 1m
  - sliding_window_moving_average:
      window_size: 10 # <---- 10 datapoints, which now also mean 10 minutes
```

---

The advantage of using resample in this case is that it is easy to understand the effect that resampling has on an irregular sensor, whereas the behaviour of the sliding time-window average I have in mind will end up being a bit mysterious.

In other words, after considering this feature I decided against implementing it, and suggest instead you use resampling before the averaging filter.

Thank you anyways for the request! It was a good idea, but it ended up not being necessary :)
